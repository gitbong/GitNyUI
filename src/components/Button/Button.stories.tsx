import React from 'react';
import { FiAperture } from 'react-icons/fi';
// import { IoIosAddCircleOutline } from 'react-icons/io';

import { Button, Theme, defaultTheme } from '../../index';

export default { title: 'Button' };

const Space = ({ children }: { children: React.ReactNode }) => (
  <div style={{ margin: '10px' }}>{children}</div>
);

export const withEmoji = () => (
  <Theme theme={defaultTheme}>
    <Space>
      <Button
        label="Primary"
        variant="primary"
        prepend={<FiAperture />}
        disabled={true}
        type="submit"
        aria-label="test button"
        tabIndex={1}
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
    <Space>
      <Button
        size="xl"
        label="Primary"
        variant="primary"
        type="submit"
        prepend={<FiAperture />}
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
    <Space>
      <Button
        label="Primary"
        variant="primary"
        outline={true}
        type="submit"
        prepend={<FiAperture />}
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
    <Space>
      <Button
        label="Secondary"
        size="md"
        variant="secondary"
        append={<FiAperture />}
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
    <Space>
      <Button
        label="Secondary"
        variant="secondary"
        outline={true}
        append={<FiAperture />}
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
    <Space>
      <Button
        label="Link"
        variant="link"
        onClick={() => {
          // alert('click');
        }}
      />
    </Space>
  </Theme>
);
