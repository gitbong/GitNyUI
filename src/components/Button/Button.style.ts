import styled, { css } from 'styled-components';
import { math, darken } from 'polished';

export const Button = styled.button<any>`
  margin:0;
  transition: 0.3s ease-out;
  box-sizing: border-box;
  appearance: button;
  border-style: solid;
  border-width: ${({ theme }) => theme.border.width};
  border-color: ${({ theme }) => theme.border.color};
  border-radius: ${({ theme }) => theme.border.radius};
  outline: none; display: flex;
  flex-direction: row;
  align-items: center;
  ${variants};
  ${size};
  ${outline};
  ${disabled};
  &:focus {
    z-index: 0;
    box-shadow: 0 0 0 4px rgba(255, 0, 0, 0.5);
  }
`;
export const PrependWrapper = styled.span`
  display: flex;
  margin-right: 0.5em;
`;
export const AppendWrapper = styled.span`
  display: flex;
  margin-left: 0.5em;
`;

function disabled({ disabled }: { disabled: boolean }) {
  if (disabled) {
    return css`
      cursor: not-allowed;
      opacity: 0.65;
    `;
  } else {
    return css`
      cursor: pointer;
    `;
  }
}

function outline({
  outline,
  variant,
  theme
}: {
  outline: boolean;
  variant: string;
  theme: any;
}) {
  let borderWidth: string;
  if (parseInt(theme.border.width) === 0) {
    borderWidth = '1px';
  } else {
    borderWidth = theme.border.width;
  }

  if (outline) {
    return css`
      background: none;
      color: ${({ theme }) => theme.color[variant].background};
      border-width: ${borderWidth};
      border-color: ${({ theme }) => theme.color[variant].background};
      &:hover {
        background-color: ${({ theme }) => theme.color[variant].background};
        color: ${({ theme }) => theme.color[variant].font};
        border-color: ${({ theme }) => theme.border.color};
      }
      &:active {
        color: ${({ theme }) => theme.color[variant].font};
        background-color: ${({ theme }) =>
          darken(0.15, `${theme.color[variant].background}`)};
      }
    `;
  }
}

function size({ size }: { size: string }): any {
  return css`
    font-size: ${({ theme }) => theme.size[size]};
    height: ${({ theme }) => math(`${theme.size[size]} * 2.5`)};
    padding-left: ${({ theme }) => math(`${theme.size[size]} * 1`)};
    padding-right: ${({ theme }) => math(`${theme.size[size]} * 1`)};
  `;
}

function variants({ variant }: { variant: string }): any {
  return css`
    color: ${({ theme }) => theme.color[variant].font};
    background-color: ${({ theme }) => theme.color[variant].background};
    &:hover {
      background-color: ${({ theme }) =>
        darken(0.05, `${theme.color[variant].background}`)};
    }
    &:active {
      background-color: ${({ theme }) =>
        darken(0.15, `${theme.color[variant].background}`)};
    }
  `;
}
