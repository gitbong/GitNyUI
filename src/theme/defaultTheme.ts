export const defaultTheme = {
  color: {
    primary: { background: '#EE3E37', font: 'white' },
    secondary: { background: '#19A9AA', font: 'white' },
    link: { background: 'white', font: 'black' }
  },
  size: {
    xs: '10px',
    sm: '12px',
    md: ' 16px',
    lg: ' 18px',
    xl: ' 20px'
  },
  space: {
    sm: { x: '10px', y: '15px' },
    md: { x: '15px', y: '20px' },
    lg: { x: '20px', y: '25px' }
  },
  border: {
    width: '5px',
    radius: '0px',
    color: 'rgba(0,0,0,1.3)'
  }
};
