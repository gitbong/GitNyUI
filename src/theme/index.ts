export * from './Theme';
export * from './defaultTheme';
