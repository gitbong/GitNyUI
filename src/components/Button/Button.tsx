import React, { MouseEvent, ReactChild } from 'react';
import * as S from './Button.style';
import { BaseHtmlUI } from '../../types';

export interface ButtonProps extends BaseHtmlUI {
  label: string;
  variant?: string;
  outline?: boolean;
  disabled?: boolean;
  prepend?: ReactChild;
  append?: ReactChild;
  onClick(e: MouseEvent): void;
}

export const Button: React.FunctionComponent<ButtonProps> = ({
  label,
  variant,
  outline,
  disabled,
  prepend,
  append,
  onClick,
  ...restProps
}) => {
  return (
    <S.Button
      variant={variant}
      outline={outline}
      disabled={disabled}
      onClick={onClick}
      {...restProps}
    >
      {prepend && <S.PrependWrapper>{prepend}</S.PrependWrapper>}
      <span>{label}</span>
      {append && <S.AppendWrapper>{append}</S.AppendWrapper>}
    </S.Button>
  );
};

Button.defaultProps = {
  variant: 'primary',
  size: 'md',
  disabled: false,
  outline: false
};
