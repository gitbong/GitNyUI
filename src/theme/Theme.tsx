import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from './defaultTheme';

interface ThemeProps {
  theme?: any;
  children: React.ReactNode;
}

export const Theme: React.FunctionComponent<ThemeProps> = ({
  theme,
  children
}) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Theme.defaultProps = {
  theme: defaultTheme
};
