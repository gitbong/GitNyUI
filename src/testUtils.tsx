import React from 'react';
import { render } from '@testing-library/react';
import { Theme, defaultTheme } from './theme';

export function renderWithTheme(component: any, theme: any = defaultTheme) {
  return render(<Theme theme={theme}>{component}</Theme>);
}
