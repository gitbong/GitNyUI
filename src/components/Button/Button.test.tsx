import React from 'react';
import userEvent from '@testing-library/user-event';
import { FiAperture } from 'react-icons/fi';
import { Button, ButtonProps } from './Button';
import { renderWithTheme } from '../../testUtils';
import { defaultTheme } from '../../theme/defaultTheme';

describe('Button', () => {
  let props: ButtonProps;
  let onClickFunction: any;

  beforeEach(() => {
    onClickFunction = jest.fn();
    props = {
      label: 'button',
      onClick: onClickFunction
    };
  });

  it('should render correctly', () => {
    const { getByText } = renderWithTheme(<Button {...props} />);
    expect(getByText('button')).toBeInTheDocument();
  });
  it('should call onClick function', () => {
    const { getByText } = renderWithTheme(<Button {...props} />);
    userEvent.click(getByText('button'));
    expect(onClickFunction).toHaveBeenCalled();
  });
  it('should not call onClick function when disabled is true', () => {
    props.disabled = true;
    const { getByText } = renderWithTheme(<Button {...props} />);
    userEvent.click(getByText('button'));
    expect(onClickFunction).not.toHaveBeenCalled();
  });

  describe('render variant', () => {
    it('should render correctly when variant is primary', () => {
      props.variant = 'primary';
      const { container } = renderWithTheme(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('render size', () => {
    it('should render correctly when size is xs', () => {
      props.size = 'xs';
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('font-size: 10px');
    });
    it('should render correctly when size is sm', () => {
      props.size = 'sm';
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('font-size: 12px');
    });
    it('should render correctly when size is md', () => {
      props.size = 'md';
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('font-size: 16px');
    });
    it('should render correctly when size is lg', () => {
      props.size = 'lg';
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('font-size: 18px');
    });
    it('should render correctly when size is xl', () => {
      props.size = 'xl';
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toHaveStyle('font-size: 20px');
    });
  });

  describe('render outline', () => {
    it('should render correctly', () => {
      props.outline = true;
      const { container } = renderWithTheme(<Button {...props} />);
      const button = container.querySelector('button');
      expect(button).toMatchSnapshot();
    });
    it('should render 1px border when theme.border.width is 0', () => {
      props.outline = true;
      const { container } = renderWithTheme(<Button {...props} />, {
        ...defaultTheme,
        border: { ...defaultTheme.border, width: '0' }
      });
      const button = container.querySelector('button');
      expect(button).toHaveStyle('border-width: 1px');
    });
  });

  describe('render prepend and append', () => {
    it('should render prepend correctly', () => {
      props.prepend = <FiAperture />;
      const { container } = renderWithTheme(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });
    it('should render append correctly', () => {
      props.append = <FiAperture />;
      const { container } = renderWithTheme(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });
  });
});
