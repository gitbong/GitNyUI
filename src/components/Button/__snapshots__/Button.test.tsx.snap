// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render outline should render 1px border when theme.border.width is 0 1`] = `
<button
  class="sc-bdVaJa dJnTLU"
>
  <span>
    button
  </span>
</button>
`;

exports[`Button render outline should render correctly 1`] = `
<button
  class="sc-bdVaJa dlDaRH"
>
  <span>
    button
  </span>
</button>
`;

exports[`Button render prepend and append should render append correctly 1`] = `
<div>
  <button
    class="sc-bdVaJa jMJazo"
  >
    <span>
      button
    </span>
    <span
      class="sc-htpNat htFjFE"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <line
          x1="14.31"
          x2="20.05"
          y1="8"
          y2="17.94"
        />
        <line
          x1="9.69"
          x2="21.17"
          y1="8"
          y2="8"
        />
        <line
          x1="7.38"
          x2="13.12"
          y1="12"
          y2="2.06"
        />
        <line
          x1="9.69"
          x2="3.95"
          y1="16"
          y2="6.06"
        />
        <line
          x1="14.31"
          x2="2.83"
          y1="16"
          y2="16"
        />
        <line
          x1="16.62"
          x2="10.88"
          y1="12"
          y2="21.94"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Button render prepend and append should render prepend correctly 1`] = `
<div>
  <button
    class="sc-bdVaJa jMJazo"
  >
    <span
      class="sc-bwzfXH eWaHkS"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <line
          x1="14.31"
          x2="20.05"
          y1="8"
          y2="17.94"
        />
        <line
          x1="9.69"
          x2="21.17"
          y1="8"
          y2="8"
        />
        <line
          x1="7.38"
          x2="13.12"
          y1="12"
          y2="2.06"
        />
        <line
          x1="9.69"
          x2="3.95"
          y1="16"
          y2="6.06"
        />
        <line
          x1="14.31"
          x2="2.83"
          y1="16"
          y2="16"
        />
        <line
          x1="16.62"
          x2="10.88"
          y1="12"
          y2="21.94"
        />
      </svg>
    </span>
    <span>
      button
    </span>
  </button>
</div>
`;

exports[`Button render variant should render correctly when variant is primary 1`] = `
<div>
  <button
    class="sc-bdVaJa jMJazo"
  >
    <span>
      button
    </span>
  </button>
</div>
`;
